src/kmeans.jl

```diff
@@ -1,91 +1,132 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+# NOTE REFACT: there's no need for this to be a mutable struct.
+
+struct KmeansResult{T<:AbstractFloat} <: ClusteringResult
+    centers::Matrix{T}        # cluster centers (d x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{T}          # cost of the assignments (n)
+    counts::Vector{Int}       # number of samples assigned to each cluster (k)
+    cweights::Vector{T}       # cluster weights (k)
+    totalcost::T              # total cost (i.e. objective)
+    iterations::Int           # number of elapsed iterations
+    converged::Bool           # whether the procedure converged
 end
 
 const _kmeans_default_init = :kmpp
 const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
+
+####
+#### Exported kmeans and kmeans! functions
+####
+
+# Case when TX is a Real but not a AbstractFloat
+function kmeans!(X::AbstractMatrix{<:Union{Integer, Rational, AbstractIrrational}},
+                 centers::AbstractMatrix{<:Real}; args...)
+    # NOTE REFACT: a warning could be issued here
+    kmeans!(convert(Matrix{Float64}, X), centers; args...)
+end
+
+# case when TX<:AbstractFloat
+function kmeans!(X::AbstractMatrix{TX}, centers::AbstractMatrix{TC};
                  weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
+                 maxiter::Int=_kmeans_default_maxiter,
```

Comment:
I don't see why `maxiter` has to be `Int`. What's wrong with e.g. `0x32`?

Reply:
I think that's just a gimmick to be honest...

Reply:
(and also that's what's used in `fuzzy_cmeans` for instance)

Reply:
Well, we started by allowing `AbstractMatrix`, and now we are disallowing innocent integral types that could be converted to `Int` with zero overhead :)
This package was started in 2014. Since then Julia evolved quite a bit, so not all of the code here follows the current idiomatic Julia (including `fuzzy_cmeans()`).
Essentially, my hope is that your PRs would improve this situation. :)

Reply:
I'm fine either way and happy to revert if you think that's better though I can't think of any practical application where one would feed the max number of iterations in another type than `Int`... it's just completely pointless.

Reply:
The question should rather be: what's the point in not accepting any `Integer`?

Reply:
🤷‍♂️ as I said, happy to revert.